pyfifotax/utils.py

```diff
@@ -0,0 +1,548 @@
+
+    total_dividends = sum_decimal(df_dividends["Amount [EUR]"])
+    total_fees = sum_decimal(df_fees["Amount [EUR]"])
+    total_taxes = sum_decimal(df_taxes["Amount [EUR]"])
+
+    # unlike a previous version, we have to split things here
+    # losses from share can only be compared to gains from shares
+    # hence, there is a "total" gain/loss and then separate gains and
+    # losses from shares, e.g. one could have had a gain of 100
+    # and a loss of 150 over a year, i.e. a total loss of 50
+    total_foreign_gains = share_losses + share_gains + total_dividends
+    gains_from_shares = share_gains
+    losses_from_shares = -share_losses
+    total_gain_forex = sum_decimal(forex_gain_series)
+
+    anlagen = [
+        (
+            "Anlage KAP",
+            "Zeile 19: Ausländische Kapitalerträge (ohne Betrag lt. Zeile 47)",
+            round_decimal(total_foreign_gains),
+        ),
+        (
+            "Anlage KAP",
+            "Zeile 20: In den Zeilen 18 und 19 enthaltene Gewinne aus Aktienveräußerungen i. S. d. § 20 Abs. 2 Satz 1 Nr 1 EStG",
+            round_decimal(gains_from_shares),
+        ),
+        (
+            "Anlage KAP",
+            "Zeile 23: In den Zeilen 18 und 19 enthaltene Verluste aus der Veräuerung von Aktien i. S. d. § 20 Abs. 2 Satz 1 Nr. 1 EStG",
+            round_decimal(losses_from_shares),
+        ),
+        (
+            "Anlage KAP",
+            "Zeile 41: Anrechenbare noch nicht angerechnete ausländische Steuern",
+            round_decimal(total_taxes),
+        ),
+        (
+            "Anlage N",
+            "Zeile 48: (Werbungskosten Sonstiges): Überweisungsgebühren auf deutsches Konto für Gehaltsbestandteil RSU/ESPP",
+            round_decimal(total_fees),
+        ),
+        (
+            "Anlage SO",
+            "Zeilen 42 - 48: Gewinn / Verlust aus Verkauf von Fremdwährungen",
+            round_decimal(total_gain_forex),  # here: the sum should be fine
+        ),
+    ]
+    summary = {
+        "ELSTER - Anlage": [a[0] for a in anlagen],
+        "ELSTER - Zeile (Suggestion!)": [a[1] for a in anlagen],
+        "Value": [a[2] for a in anlagen],
+    }
+    df_summary = pd.DataFrame(summary)
+    return df_summary
+
+
+def create_report_sheet(name: str, df: pd.DataFrame, writer: pd.ExcelWriter):
+    if df.empty:
+        return
+
+    if "Date" in df:
+        df.sort_values("Date", inplace=True)
+    elif "date" in df:
+        df.sort_values("date", inplace=True)
+    elif "Sell Date" in df:
+        df.sort_values(["Sell Date", "Buy Date"], inplace=True)
+    elif "Meldezeitraum" in df:
+        df.sort_values("Meldezeitraum", inplace=True)
+    elif "ELSTER" in df.columns[0]:
+        pass  # don't sort ELSTER frames
+    else:
+        raise RuntimeError(
+            "Couldn't sort data, expected either 'Date', 'date', 'Sell Date', or 'Meldezeitraum' column to exist"
+        )
+
+    df.to_excel(writer, sheet_name=name, index=False)
```

Comment:
It seems that in the generated report the numbers are actually strings (probably because of the `Decimal` representation).

These can hopefully fix it:
- `df.apply(pd.to_numeric)` with `float_format="%.2f"` in the `to_excel` function
- `df.column_name = df.column_name.astype(float)`

Reply:
Indeed. I'll convert Decimals into floats before writing, good point.